cpp/src/structure/detail/structure_utils.cuh

```diff
@@ -496,6 +498,62 @@ void sort_adjacency_list(raft::handle_t const& handle,
   }
 }
 
+template <typename T>
+struct indirect_array_reference {
+  T const* array_;
+
+  T operator() __host__ __device__(size_t index) { return array_[index]; }
+};
+
+template <typename vertex_t, typename comparison_t>
+std::tuple<size_t, rmm::device_uvector<uint32_t>> mark_edges_for_removal(
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> src,
+  raft::device_span<vertex_t const> dst,
+  comparison_t comparison)
+{
+  rmm::device_uvector<uint32_t> remove_flags(packed_bool_size(src.size()), handle.get_stream());
+  thrust::fill(handle.get_thrust_policy(),
+               remove_flags.begin(),
+               remove_flags.end(),
+               cugraph::packed_bool_empty_mask());
+
+  size_t remove_count = thrust::count_if(
+    handle.get_thrust_policy(),
+    thrust::make_counting_iterator(size_t{0}),
+    thrust::make_counting_iterator(src.size()),
+    [comparison, d_remove_flags = remove_flags.data()] __device__(size_t i) {
+      if (comparison(i)) {
+        atomicOr(d_remove_flags + cugraph::packed_bool_offset(i), cugraph::packed_bool_mask(i));
+        return true;
+      }
+      return false;
+    });
```

Comment:
Should we better call this `remove_flags.size()`? We can avoid atomics and no need to call thrust::fill.

Something like
```
thrust::tabulate(
  handle.get_thrust_policy(),
  remove_flags.begin(),
  remove_flags.end(),
  [edge_first, comparison]__device__(auto i) {
    auto word = packed_bool_empty_mask();
    ...
    return word;
  });
```

Reply:
I can rework this.